pkg/server/chaosd/kafka.go

```diff
@@ -0,0 +1,381 @@
+
+func (j kafkaAttack) Attack(options core.AttackConfig, env Environment) (err error) {
+	attack := options.(*core.KafkaCommand)
+	switch attack.Action {
+	case core.KafkaFillAction:
+		return attackKafkaFill(context.TODO(), attack)
+	case core.KafkaFloodAction:
+		return attackKafkaFlood(context.TODO(), attack)
+	case core.KafkaIOAction:
+		return attackKafkaIO(attack)
+	default:
+		return errors.Errorf("invalid action: %s", attack.Action)
+	}
+}
+
+func (j kafkaAttack) Recover(exp core.Experiment, env Environment) error {
+	attack := new(core.KafkaCommand)
+	if err := json.Unmarshal([]byte(exp.RecoverCommand), attack); err != nil {
+		return perr.Wrap(err, "unmarshal kafka command")
+	}
+	if attack.Action == core.KafkaIOAction {
+		return recoverKafkaIO(attack, env)
+	}
+	return nil
+}
+
+func newDialer(attack *core.KafkaCommand) (dialer *client.Dialer, err error) {
+	dialer = &client.Dialer{
+		Timeout:   10 * time.Second,
+		DualStack: true,
+	}
+	if attack.Username != "" {
+		dialer.SASLMechanism, err = scram.Mechanism(scram.SHA512, attack.Username, attack.Password)
+		if err != nil {
+			return nil, perr.Wrap(err, "create scram mechanism")
+		}
+	}
+	return dialer, nil
+}
+
+func dial(attack *core.KafkaCommand) (conn *client.Conn, err error) {
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	dialer, err := newDialer(attack)
+	if err != nil {
+		return nil, perr.Wrap(err, "new dialer")
+	}
+	conn, err = dialer.Dial("tcp", endpoint)
+	if err != nil {
+		return nil, perr.Wrapf(err, "dial endpoint: %s", endpoint)
+	}
+	return conn, nil
+}
+
+func dialPartition(ctx context.Context, attack *core.KafkaCommand, partition int) (conn *client.Conn, err error) {
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	dialer, err := newDialer(attack)
+	if err != nil {
+		return nil, perr.Wrap(err, "new dialer")
+	}
+	conn, err = dialer.DialLeader(ctx, "tcp", endpoint, attack.Topic, partition)
+	if err != nil {
+		return nil, perr.Wrapf(err, "dial endpoint: %s", endpoint)
+	}
+	return conn, nil
+}
+
+func getPartitions(attack *core.KafkaCommand) (partitions []int, err error) {
+	dialer, err := newDialer(attack)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := fmt.Sprintf("%s:%d", attack.Host, attack.Port)
+	conn, err := dialer.Dial("tcp", endpoint)
+	if err != nil {
+		return nil, errors.Wrapf(err, "dial endpoint: %s", endpoint)
+	}
+
+	if attack.Topic == "" {
+		return nil, errors.New("topic is required")
+	}
+	pars, err := conn.ReadPartitions(attack.Topic)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read partitions of topic %s", attack.Topic)
+	}
+	for _, par := range pars {
+		if par.Error != nil {
+			return nil, errors.Wrap(err, "read partition")
+		}
+		partitions = append(partitions, par.ID)
+	}
+	return partitions, nil
+}
+
+func attackKafkaFill(ctx context.Context, attack *core.KafkaCommand) (err error) {
+	// TODO: make it configurable
+	const messagePerRequest = 128
+	conn, err := dial(attack)
```

Comment:
Do we need to use a conn pool here? I see that only one conn is actually created, but the following is to use multiple goroutines to send messages.

Reply:
Oh, my fault, but I found this `conn` was not used in the following logic.